### PR TITLE
Removing MSVC 2013 conditional code from C++ code

### PR DESCRIPTION
**Pull request by bhardwajs**
In PR #1030, we remove the support for MSVC 2013 from Bond CI pipeline. PR #1032 cleans up compiler of MSVC 2013 conditional code. This PR removes MSVC 2013 conditional code from C++ and simplifies the resulting C++ code.

This PR can only be merged after #1030 is merged.

### CODE REVIEW

File: cpp/test/core/maybe_tests.cpp [hunk 1]
```diff
@@ -377,7 +357,6 @@ TEST_CASE_BEGIN(AllocatorPropagated)
 }
 TEST_CASE_END
 
-#if !defined(_MSC_VER) || _MSC_VER >= 1900
 // MSVC 2013 doesn't have implementations of these type traits, so we skip
```
Comment (chwarr) [inline cpp/test/core/maybe_tests.cpp:360]: Minor: this comment looks stale

File: examples/cpp/core/scoped_allocator/allocator.h [hunk 2]
```diff
@@ -12,12 +12,6 @@ namespace scoped_allocator
     {
         using value_type = T;
 
-#if defined(_MSC_VER) && _MSC_VER < 1910
-        // MSVC 2015 and older versions have bug in std::scoped_allocator_adaptor
```
Comment (chwarr) [inline examples/cpp/core/scoped_allocator/allocator.h:16]: We're still targeting MSVC 2015, so I think this need to stay.

Comment (bhardwajs) [reply examples/cpp/core/scoped_allocator/allocator.h:16]: Yeah, good catch. Thanks!

**Comment by bhardwajs** [comment]
@chwarr - Thanks for the review. I have rebased to maser so build should succeed.
I have removed `inc/bond/core/warning.h` and associated code block in config.cmake.
Also took care of the two minor comments.

**Pull request merged**